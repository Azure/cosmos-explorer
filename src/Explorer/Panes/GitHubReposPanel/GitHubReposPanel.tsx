import React from "react";
import { Areas, HttpStatusCodes } from "../../../Common/Constants";
import { handleError } from "../../../Common/ErrorHandlingUtils";
import { GitHubClient, IGitHubPageInfo, IGitHubRepo } from "../../../GitHub/GitHubClient";
import { useSidePanel } from "../../../hooks/useSidePanel";
import { IPinnedRepo, JunoClient } from "../../../Juno/JunoClient";
import { Action, ActionModifiers } from "../../../Shared/Telemetry/TelemetryConstants";
import * as TelemetryProcessor from "../../../Shared/Telemetry/TelemetryProcessor";
import * as GitHubUtils from "../../../Utils/GitHubUtils";
import * as JunoUtils from "../../../Utils/JunoUtils";
import { AuthorizeAccessComponent } from "../../Controls/GitHub/AuthorizeAccessComponent";
import {
  GitHubReposComponent,
  GitHubReposComponentProps,
  RepoListItem,
} from "../../Controls/GitHub/GitHubReposComponent";
import { ContentMainStyle } from "../../Controls/GitHub/GitHubStyleConstants";
import { BranchesProps, PinnedReposProps, UnpinnedReposProps } from "../../Controls/GitHub/ReposListComponent";
import Explorer from "../../Explorer";
import { PanelInfoErrorComponent } from "../PanelInfoErrorComponent";
import { PanelLoadingScreen } from "../PanelLoadingScreen";

interface IGitHubReposPanelProps {
  explorer: Explorer;
  gitHubClientProp: GitHubClient;
  junoClientProp?: JunoClient;
}

interface IGitHubReposPanelState {
  showAuthorizationAcessState: boolean;
  isExecuting: boolean;
  errorMessage: string;
  showErrorDetails: boolean;
  gitHubReposState: GitHubReposComponentProps;
}
export class GitHubReposPanel extends React.Component<IGitHubReposPanelProps, IGitHubReposPanelState> {
  private static readonly PageSize = 30;
  private isAddedRepo = false;
  private gitHubClient: GitHubClient;
  private junoClient: JunoClient;

  private branchesProps: Record<string, BranchesProps>;
  private pinnedReposProps: PinnedReposProps;
  private unpinnedReposProps: UnpinnedReposProps;

  private allGitHubRepos: IGitHubRepo[];
  private allGitHubReposLastPageInfo?: IGitHubPageInfo;
  private pinnedReposUpdated: boolean;

  constructor(props: IGitHubReposPanelProps) {
    super(props);

    this.unpinnedReposProps = {
      repos: [],
      hasMore: true,
      isLoading: true,
      loadMore: (): Promise<void> => this.loadMoreUnpinnedRepos(),
    };
    this.branchesProps = {};
    this.pinnedReposProps = {
      repos: [],
    };

    this.allGitHubRepos = [];
    this.allGitHubReposLastPageInfo = undefined;
    this.pinnedReposUpdated = false;

    this.state = {
      showAuthorizationAcessState: true,
      isExecuting: false,
      errorMessage: "",
      showErrorDetails: false,
      gitHubReposState: {
        showAuthorizeAccess: !this.props.explorer.notebookManager?.gitHubOAuthService.isLoggedIn(),
        authorizeAccessProps: {
          scope: this.getOAuthScope(),
          authorizeAccess: (scope): void => this.connectToGitHub(scope),
        },
        reposListProps: {
          branchesProps: this.branchesProps,
          pinnedReposProps: this.pinnedReposProps,
          unpinnedReposProps: this.unpinnedReposProps,
          pinRepo: (item): Promise<void> => this.pinRepo(item),
          unpinRepo: (item): Promise<void> => this.unpinRepo(item),
        },
        addRepoProps: {
          container: this.props.explorer,
          getRepo: (owner, repo): Promise<IGitHubRepo> => this.getRepo(owner, repo),
          pinRepo: (item): Promise<void> => this.pinRepo(item),
        },
        resetConnection: (): void => this.setup(true),
        onOkClick: (): Promise<void> => this.submit(),
        onCancelClick: (): void => useSidePanel.getState().closeSidePanel(),
      },
    };
    this.gitHubClient = this.props.gitHubClientProp;
    this.junoClient = this.props.junoClientProp;
  }

  componentDidMount(): void {
    this.open();
  }

  public open(): void {
    this.resetData();
    this.setup();
  }

  public async submit(): Promise<void> {
    const pinnedReposUpdated = this.pinnedReposUpdated;
    const reposToPin: IPinnedRepo[] = this.pinnedReposProps.repos.map((repo) => JunoUtils.toPinnedRepo(repo));

    if (pinnedReposUpdated) {
      try {
        const response = await this.junoClient.updatePinnedRepos(reposToPin);
        if (response.status !== HttpStatusCodes.OK) {
          throw new Error(`Received HTTP ${response.status} when saving pinned repos`);
        }
      } catch (error) {
        handleError(error, "GitHubReposPane/submit", "Failed to save pinned repos");
      }
    }
    useSidePanel.getState().closeSidePanel();
  }

  public resetData(): void {
    this.branchesProps = {};

    this.pinnedReposProps.repos = [];
    this.unpinnedReposProps.repos = [];
    this.allGitHubRepos = [];
    this.allGitHubReposLastPageInfo = undefined;

    this.pinnedReposUpdated = false;
    this.unpinnedReposProps.hasMore = true;
    this.unpinnedReposProps.isLoading = true;
  }

  private getOAuthScope(): string {
    return (
      this.props.explorer.notebookManager?.gitHubOAuthService.getTokenObservable()()?.scope ||
      AuthorizeAccessComponent.Scopes.Public.key
    );
  }

  private setup(forceShowConnectToGitHub = false): void {
    forceShowConnectToGitHub || !this.props.explorer.notebookManager?.gitHubOAuthService.isLoggedIn()
      ? this.setupForConnectToGitHub(forceShowConnectToGitHub)
      : this.setupForManageRepos();
  }

  private setupForConnectToGitHub(forceShowConnectToGitHub: boolean): void {
    if (forceShowConnectToGitHub) {
      const newState = { ...this.state.gitHubReposState };
      newState.showAuthorizeAccess = forceShowConnectToGitHub;
      this.setState({
        gitHubReposState: newState,
      });
    }
    this.setState({
      isExecuting: false,
    });
  }

  private async setupForManageRepos(): Promise<void> {
    this.setState({
      isExecuting: false,
    });
    TelemetryProcessor.trace(Action.NotebooksGitHubManageRepo, ActionModifiers.Mark, {
      dataExplorerArea: Areas.Notebook,
    });

    this.refreshManageReposComponent();
  }

  private calculateUnpinnedRepos(): RepoListItem[] {
    const unpinnedGitHubRepos = this.allGitHubRepos.filter(
      (gitHubRepo) =>
        this.pinnedReposProps.repos.findIndex(
          (pinnedRepo) => pinnedRepo.key === GitHubUtils.toRepoFullName(gitHubRepo.owner, gitHubRepo.name)
        ) === -1
    );
    return unpinnedGitHubRepos.map((gitHubRepo) => ({
      key: GitHubUtils.toRepoFullName(gitHubRepo.owner, gitHubRepo.name),
      repo: gitHubRepo,
      branches: [],
    }));
  }

  private async loadMoreBranches(repo: IGitHubRepo): Promise<void> {
    const branchesProps = this.branchesProps[GitHubUtils.toRepoFullName(repo.owner, repo.name)];
    branchesProps.hasMore = true;
    branchesProps.isLoading = true;

    try {
      const response = await this.gitHubClient.getBranchesAsync(
        repo.owner,
        repo.name,
        GitHubReposPanel.PageSize,
        branchesProps.lastPageInfo?.endCursor
      );

      if (response.status !== HttpStatusCodes.OK) {
        throw new Error(`Received HTTP ${response.status} when fetching branches`);
      }

      if (response.data) {
        branchesProps.branches = branchesProps.branches.concat(response.data);
        branchesProps.lastPageInfo = response.pageInfo;
      }
    } catch (error) {
      handleError(error, "GitHubReposPane/loadMoreBranches", "Failed to fetch branches");
    }

    branchesProps.isLoading = false;
    branchesProps.hasMore = branchesProps.lastPageInfo?.hasNextPage;
    this.setState({
      gitHubReposState: {
        ...this.state.gitHubReposState,
        reposListProps: {
          ...this.state.gitHubReposState.reposListProps,
          branchesProps: {
            ...this.state.gitHubReposState.reposListProps.branchesProps,
            [GitHubUtils.toRepoFullName(repo.owner, repo.name)]: branchesProps,
          },
          pinnedReposProps: {
            repos: this.pinnedReposProps.repos,
          },
          unpinnedReposProps: {
            ...this.state.gitHubReposState.reposListProps.unpinnedReposProps,
            repos: this.unpinnedReposProps.repos,
          },
        },
      },
    });
  }

  private async loadMoreUnpinnedRepos(): Promise<void> {
    this.unpinnedReposProps.isLoading = true;
    this.unpinnedReposProps.hasMore = true;

    try {
      const response = await this.gitHubClient.getReposAsync(
        GitHubReposPanel.PageSize,
        this.allGitHubReposLastPageInfo?.endCursor
      );

      if (response.status !== HttpStatusCodes.OK) {
        throw new Error(`Received HTTP ${response.status} when fetching unpinned repos`);
      }

      if (response.data) {
        this.allGitHubRepos = this.allGitHubRepos.concat(response.data);
        this.allGitHubReposLastPageInfo = response.pageInfo;
        this.unpinnedReposProps.repos = this.calculateUnpinnedRepos();
      }
    } catch (error) {
      handleError(error, "GitHubReposPane/loadMoreUnpinnedRepos", "Failed to fetch unpinned repos");
    }

    this.unpinnedReposProps.isLoading = false;
    this.unpinnedReposProps.hasMore = this.allGitHubReposLastPageInfo?.hasNextPage;

    this.setState({
      gitHubReposState: {
        ...this.state.gitHubReposState,
        reposListProps: {
          ...this.state.gitHubReposState.reposListProps,
          unpinnedReposProps: {
            ...this.state.gitHubReposState.reposListProps.unpinnedReposProps,
            isLoading: this.unpinnedReposProps.isLoading,
            hasMore: this.unpinnedReposProps.hasMore,
            repos: this.unpinnedReposProps.repos,
          },
        },
      },
    });
  }

  private async getRepo(owner: string, repo: string): Promise<IGitHubRepo> {
    try {
      const response = await this.gitHubClient.getRepoAsync(owner, repo);
      if (response.status !== HttpStatusCodes.OK) {
        throw new Error(`Received HTTP ${response.status} when fetching repo`);
      }
      this.isAddedRepo = true;
      return response.data;
    } catch (error) {
      handleError(error, "GitHubReposPane/getRepo", "Failed to fetch repo");
      return Promise.resolve(undefined);
    }
  }

  private async pinRepo(item: RepoListItem): Promise<void> {
    this.pinnedReposUpdated = true;
    const initialReposLength = this.pinnedReposProps.repos.length;

    const existingRepo = this.pinnedReposProps.repos.find((repo) => repo.key === item.key);
    if (existingRepo) {
      existingRepo.branches = item.branches;
    } else {
      this.pinnedReposProps.repos = [...this.pinnedReposProps.repos, item];
    }

    this.unpinnedReposProps.repos = this.calculateUnpinnedRepos();

    if (this.pinnedReposProps.repos.length > initialReposLength) {
      this.refreshBranchesForPinnedRepos();
    }
  }

  private async unpinRepo(item: RepoListItem): Promise<void> {
    this.pinnedReposUpdated = true;
    this.pinnedReposProps.repos = this.pinnedReposProps.repos.filter((pinnedRepo) => pinnedRepo.key !== item.key);
    this.unpinnedReposProps.repos = this.calculateUnpinnedRepos();

    this.setState({
      gitHubReposState: {
        ...this.state.gitHubReposState,
        reposListProps: {
          ...this.state.gitHubReposState.reposListProps,
          pinnedReposProps: {
            repos: this.pinnedReposProps.repos,
          },
          unpinnedReposProps: {
            ...this.state.gitHubReposState.reposListProps.unpinnedReposProps,
            repos: this.unpinnedReposProps.repos,
          },
        },
      },
    });
  }

  private async refreshManageReposComponent(): Promise<void> {
    await this.refreshPinnedRepoListItems();
    this.refreshBranchesForPinnedRepos();
    this.refreshUnpinnedRepoListItems();
  }

  private async refreshPinnedRepoListItems(): Promise<void> {
    this.pinnedReposProps.repos = [];

    try {
      const response = await this.junoClient.getPinnedRepos(
        this.props.explorer.notebookManager?.gitHubOAuthService.getTokenObservable()()?.scope
      );

      if (response.status !== HttpStatusCodes.OK && response.status !== HttpStatusCodes.NoContent) {
        throw new Error(`Received HTTP ${response.status} when fetching pinned repos`);
      }

      if (response.data) {
        const pinnedRepos = response.data.map(
          (pinnedRepo) =>
            ({
              key: GitHubUtils.toRepoFullName(pinnedRepo.owner, pinnedRepo.name),
              branches: pinnedRepo.branches,
              repo: JunoUtils.toGitHubRepo(pinnedRepo),
            } as RepoListItem)
        );

        this.pinnedReposProps.repos = pinnedRepos;
      }
    } catch (error) {
      handleError(error, "GitHubReposPane/refreshPinnedReposListItems", "Failed to fetch pinned repos");
    }
  }

  private refreshBranchesForPinnedRepos(): void {
    this.pinnedReposProps.repos.map((item) => {
      if (!this.branchesProps[item.key]) {
        this.branchesProps[item.key] = {
          branches: [],
          lastPageInfo: undefined,
          hasMore: true,
          isLoading: true,
          loadMore: (): Promise<void> => this.loadMoreBranches(item.repo),
        };
        this.loadMoreBranches(item.repo);
      }
    });

    this.setState({
      gitHubReposState: {
        ...this.state.gitHubReposState,
        reposListProps: {
          ...this.state.gitHubReposState.reposListProps,
          branchesProps: {
            ...this.branchesProps,
          },
          pinnedReposProps: {
            repos: this.pinnedReposProps.repos,
          },
          unpinnedReposProps: {
            ...this.state.gitHubReposState.reposListProps.unpinnedReposProps,
            repos: this.unpinnedReposProps.repos,
          },
        },
      },
    });
    this.isAddedRepo = false;
  }

  private async refreshUnpinnedRepoListItems(): Promise<void> {
    this.allGitHubRepos = [];
    this.allGitHubReposLastPageInfo = undefined;
    this.unpinnedReposProps.repos = [];

    this.loadMoreUnpinnedRepos();
  }

  private connectToGitHub(scope: string): void {
    this.setState({
      isExecuting: true,
    });
    TelemetryProcessor.trace(Action.NotebooksGitHubAuthorize, ActionModifiers.Mark, {
      dataExplorerArea: Areas.Notebook,
      scopesSelected: scope,
    });
    this.props.explorer.notebookManager?.gitHubOAuthService.startOAuth(scope);
  }

  render(): JSX.Element {
    return (
      <form className="panelFormWrapper">
        {this.state.errorMessage && (
          <PanelInfoErrorComponent
            message={this.state.errorMessage}
            messageType="error"
            showErrorDetails={this.state.showErrorDetails}
          />
        )}
        <div className="panelMainContent" style={ContentMainStyle}>
          <GitHubReposComponent {...this.state.gitHubReposState} />
        </div>

        {this.state.isExecuting && <PanelLoadingScreen />}
      </form>
    );
  }
}
