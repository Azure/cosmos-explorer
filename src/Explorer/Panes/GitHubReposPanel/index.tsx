import { useBoolean } from "@uifabric/react-hooks";
import React, { FunctionComponent, useEffect, useState } from "react";
import { Areas, HttpStatusCodes } from "../../../Common/Constants";
import { handleError } from "../../../Common/ErrorHandlingUtils";
import { GitHubClient, IGitHubPageInfo, IGitHubRepo } from "../../../GitHub/GitHubClient";
import { IPinnedRepo, JunoClient } from "../../../Juno/JunoClient";
import { Action, ActionModifiers } from "../../../Shared/Telemetry/TelemetryConstants";
import * as TelemetryProcessor from "../../../Shared/Telemetry/TelemetryProcessor";
import * as GitHubUtils from "../../../Utils/GitHubUtils";
import * as JunoUtils from "../../../Utils/JunoUtils";
import { AuthorizeAccessComponent } from "../../Controls/GitHub/AuthorizeAccessComponent";
import {
  GitHubReposComponent,
  GitHubReposComponentProps,
  RepoListItem,
} from "../../Controls/GitHub/GitHubReposComponent";
import { GitHubReposComponentAdapter } from "../../Controls/GitHub/GitHubReposComponentAdapter";
import { BranchesProps, PinnedReposProps, UnpinnedReposProps } from "../../Controls/GitHub/ReposListComponent";
import Explorer from "../../Explorer";
import { GenericRightPaneComponent, GenericRightPaneProps } from "../GenericRightPaneComponent";

// import * as ViewModels from "../../../Contracts/ViewModels";
// import { GitHubReposComponentProps, RepoListItem } from "../../Controls/GitHub/GitHubReposComponent";
interface IGitHubReposPanelProps {
  explorer: Explorer;
  closePanel: () => void;
  gitHubClientProp: GitHubClient;
  junoClientProp: JunoClient;
  panelTitle: string;
}

export const GitHubReposPanel: FunctionComponent<IGitHubReposPanelProps> = ({
  explorer,
  closePanel,
  gitHubClientProp,
  junoClientProp,
  panelTitle,
}: IGitHubReposPanelProps): JSX.Element => {
  const title = panelTitle;
  const [isLoading, { setTrue: setLoadingTrue, setFalse: setLoadingFalse }] = useBoolean(false);
  const [formError, setFormError] = useState<string>("");
  const [formErrorsDetails, setFormErrorsDetails] = useState<string>("");
  const genericPaneProps: GenericRightPaneProps = {
    container: explorer,
    formError: formError,
    formErrorDetail: formErrorsDetails,
    id: "gitHubReposPanel",
    isExecuting: isLoading,
    title: title,
    submitButtonText: "",
    onClose: () => closePanel(),
    onSubmit: undefined,
    isFooterHidden: true,
  };

  const PageSize = 30;

  const gitHubClient: GitHubClient = gitHubClientProp;
  const junoClient: JunoClient = junoClientProp;

  const branchesProps: Record<string, BranchesProps> = {};
  const pinnedReposProps: PinnedReposProps = {
    repos: [],
  };
  const unpinnedReposProps: UnpinnedReposProps = {
    repos: [],
    hasMore: true,
    isLoading: true,
    loadMore: (): Promise<void> => loadMoreUnpinnedRepos(),
  };

  const getOAuthScope = (): string => {
    return (
      explorer.notebookManager?.gitHubOAuthService.getTokenObservable()()?.scope ||
      AuthorizeAccessComponent.Scopes.Public.key
    );
  };

  const gitHubReposProps: GitHubReposComponentProps = {
    showAuthorizeAccess: true,
    authorizeAccessProps: {
      scope: getOAuthScope(),
      authorizeAccess: (scope): void => connectToGitHub(scope),
    },
    reposListProps: {
      branchesProps: branchesProps,
      pinnedReposProps: pinnedReposProps,
      unpinnedReposProps: unpinnedReposProps,
      pinRepo: (item): Promise<void> => pinRepo(item),
      unpinRepo: (item): Promise<void> => unpinRepo(item),
    },
    addRepoProps: {
      container: explorer,
      getRepo: (owner, repo): Promise<IGitHubRepo> => getRepo(owner, repo),
      pinRepo: (item): Promise<void> => pinRepo(item),
    },
    resetConnection: (): void => setup(true),
    onOkClick: (): Promise<void> => submit(),
    onCancelClick: (): void => closePanel(),
  };
  const gitHubReposAdapter: GitHubReposComponentAdapter = new GitHubReposComponentAdapter(gitHubReposProps);
  const [allGitHubRepos, setAllGitHubRepos] = useState<IGitHubRepo[]>([]);
  const [allGitHubReposLastPageInfo, setAllGitHubReposLastPageInfo] = useState<IGitHubPageInfo>(undefined);
  const [pinnedReposUpdatedState, setPinnedReposUpdated] = useState(false);

  useEffect(() => {
    resetData();
    setup();
  }, []);

  const submit = async (): Promise<void> => {
    setFormError("");
    setFormErrorsDetails("");

    const pinnedReposUpdated = pinnedReposUpdatedState;
    const reposToPin: IPinnedRepo[] = pinnedReposProps.repos.map((repo) => JunoUtils.toPinnedRepo(repo));

    if (pinnedReposUpdated) {
      try {
        const response = await junoClient.updatePinnedRepos(reposToPin);
        if (response.status !== HttpStatusCodes.OK) {
          throw new Error(`Received HTTP ${response.status} when saving pinned repos`);
        }
      } catch (error) {
        handleError(error, "GitHubReposPane/submit", "Failed to save pinned repos");
      }
    }
  };

  const resetData = (): void => {
    // Reset cached branches
    gitHubReposProps.reposListProps.branchesProps = branchesProps;

    // Reset cached pinned and unpinned repos
    pinnedReposProps.repos = [];
    unpinnedReposProps.repos = [];

    // Reset cached repos
    setAllGitHubRepos([]);
    setAllGitHubReposLastPageInfo(undefined);

    // Reset flags
    setPinnedReposUpdated(false);
    unpinnedReposProps.hasMore = true;
    unpinnedReposProps.isLoading = true;

    triggerRender();
  };

  const setup = (forceShowConnectToGitHub = false): void => {
    //eslint-disable-next-line
    console.log("setup > ", forceShowConnectToGitHub);

    forceShowConnectToGitHub || !explorer.notebookManager?.gitHubOAuthService.isLoggedIn()
      ? setupForConnectToGitHub()
      : setupForManageRepos();
  };

  const setupForConnectToGitHub = (): void => {
    //eslint-disable-next-line
    console.log("setupForConnectToGitHub");
    gitHubReposProps.showAuthorizeAccess = false;
    gitHubReposProps.authorizeAccessProps.scope = getOAuthScope();
    setLoadingFalse();
    // setTitle(GitHubReposComponent.ConnectToGitHubTitle); // Used for telemetry
    triggerRender();
  };

  const setupForManageRepos = async (): Promise<void> => {
    //eslint-disable-next-line
    console.log("setupForManageRepos");

    gitHubReposProps.showAuthorizeAccess = true;
    setLoadingFalse();
    // setTitle(GitHubReposComponent.ManageGitHubRepoTitle);
    TelemetryProcessor.trace(Action.NotebooksGitHubManageRepo, ActionModifiers.Mark, {
      dataExplorerArea: Areas.Notebook,
    });
    triggerRender();

    refreshManageReposComponent();
  };

  const calculateUnpinnedRepos = (): RepoListItem[] => {
    const unpinnedGitHubRepos = allGitHubRepos.filter(
      (gitHubRepo) =>
        pinnedReposProps.repos.findIndex(
          (pinnedRepo) => pinnedRepo.key === GitHubUtils.toRepoFullName(gitHubRepo.owner, gitHubRepo.name)
        ) === -1
    );
    return unpinnedGitHubRepos.map((gitHubRepo) => ({
      key: GitHubUtils.toRepoFullName(gitHubRepo.owner, gitHubRepo.name),
      repo: gitHubRepo,
      branches: [],
    }));
  };

  const loadMoreBranches = async (repo: IGitHubRepo): Promise<void> => {
    const branchesPropsLoadMore = branchesProps[GitHubUtils.toRepoFullName(repo.owner, repo.name)];
    branchesPropsLoadMore.hasMore = true;
    branchesPropsLoadMore.isLoading = true;
    triggerRender();

    try {
      const response = await gitHubClient.getBranchesAsync(
        repo.owner,
        repo.name,
        PageSize,
        branchesPropsLoadMore.lastPageInfo?.endCursor
      );
      if (response.status !== HttpStatusCodes.OK) {
        throw new Error(`Received HTTP ${response.status} when fetching branches`);
      }

      if (response.data) {
        branchesPropsLoadMore.branches = branchesPropsLoadMore.branches.concat(response.data);
        branchesPropsLoadMore.lastPageInfo = response.pageInfo;
      }
    } catch (error) {
      handleError(error, "GitHubReposPane/loadMoreBranches", "Failed to fetch branches");
    }

    branchesPropsLoadMore.isLoading = false;
    branchesPropsLoadMore.hasMore = branchesPropsLoadMore.lastPageInfo?.hasNextPage;
    triggerRender();
  };

  const loadMoreUnpinnedRepos = async (): Promise<void> => {
    unpinnedReposProps.isLoading = true;
    unpinnedReposProps.hasMore = true;
    triggerRender();

    try {
      const response = await gitHubClient.getReposAsync(PageSize, allGitHubReposLastPageInfo?.endCursor);
      if (response.status !== HttpStatusCodes.OK) {
        throw new Error(`Received HTTP ${response.status} when fetching unpinned repos`);
      }

      if (response.data) {
        setAllGitHubRepos(allGitHubRepos.concat(response.data));
        setAllGitHubReposLastPageInfo(response.pageInfo);
        unpinnedReposProps.repos = calculateUnpinnedRepos();
      }
    } catch (error) {
      handleError(error, "GitHubReposPane/loadMoreUnpinnedRepos", "Failed to fetch unpinned repos");
    }

    unpinnedReposProps.isLoading = false;
    unpinnedReposProps.hasMore = allGitHubReposLastPageInfo?.hasNextPage;
    triggerRender();
  };

  const getRepo = async (owner: string, repo: string): Promise<IGitHubRepo> => {
    try {
      const response = await gitHubClient.getRepoAsync(owner, repo);
      if (response.status !== HttpStatusCodes.OK) {
        throw new Error(`Received HTTP ${response.status} when fetching repo`);
      }

      return response.data;
    } catch (error) {
      handleError(error, "GitHubReposPane/getRepo", "Failed to fetch repo");
      return Promise.resolve(undefined);
    }
  };

  const pinRepo = async (item: RepoListItem): Promise<void> => {
    setPinnedReposUpdated(true);
    const initialReposLength = pinnedReposProps.repos.length;

    const existingRepo = pinnedReposProps.repos.find((repo) => repo.key === item.key);
    if (existingRepo) {
      existingRepo.branches = item.branches;
    } else {
      pinnedReposProps.repos = [...pinnedReposProps.repos, item];
    }

    unpinnedReposProps.repos = calculateUnpinnedRepos();
    triggerRender();

    if (pinnedReposProps.repos.length > initialReposLength) {
      refreshBranchesForPinnedRepos();
    }
  };

  const unpinRepo = async (item: RepoListItem): Promise<void> => {
    setPinnedReposUpdated(true);
    pinnedReposProps.repos = pinnedReposProps.repos.filter((pinnedRepo) => pinnedRepo.key !== item.key);
    unpinnedReposProps.repos = calculateUnpinnedRepos();
    triggerRender();
  };

  const refreshManageReposComponent = async (): Promise<void> => {
    await refreshPinnedRepoListItems();
    refreshBranchesForPinnedRepos();
    refreshUnpinnedRepoListItems();
  };

  const refreshPinnedRepoListItems = async (): Promise<void> => {
    pinnedReposProps.repos = [];
    triggerRender();

    try {
      const response = await junoClient.getPinnedRepos(
        explorer.notebookManager?.gitHubOAuthService.getTokenObservable()()?.scope
      );
      if (response.status !== HttpStatusCodes.OK && response.status !== HttpStatusCodes.NoContent) {
        throw new Error(`Received HTTP ${response.status} when fetching pinned repos`);
      }

      if (response.data) {
        const pinnedRepos = response.data.map(
          (pinnedRepo) =>
            ({
              key: GitHubUtils.toRepoFullName(pinnedRepo.owner, pinnedRepo.name),
              branches: pinnedRepo.branches,
              repo: JunoUtils.toGitHubRepo(pinnedRepo),
            } as RepoListItem)
        );

        pinnedReposProps.repos = pinnedRepos;
        triggerRender();
      }
    } catch (error) {
      handleError(error, "GitHubReposPane/refreshPinnedReposListItems", "Failed to fetch pinned repos");
    }
  };

  const refreshBranchesForPinnedRepos = (): void => {
    pinnedReposProps.repos.map((item) => {
      if (!branchesProps[item.key]) {
        branchesProps[item.key] = {
          branches: [],
          lastPageInfo: undefined,
          hasMore: true,
          isLoading: true,
          loadMore: (): Promise<void> => loadMoreBranches(item.repo),
        };
        loadMoreBranches(item.repo);
      }
    });
  };

  const refreshUnpinnedRepoListItems = async (): Promise<void> => {
    setAllGitHubRepos([]);
    setAllGitHubReposLastPageInfo(undefined);
    unpinnedReposProps.repos = [];
    loadMoreUnpinnedRepos();
  };

  const connectToGitHub = (scope: string): void => {
    setLoadingTrue();
    TelemetryProcessor.trace(Action.NotebooksGitHubAuthorize, ActionModifiers.Mark, {
      dataExplorerArea: Areas.Notebook,
      scopesSelected: scope,
    });
    explorer.notebookManager?.gitHubOAuthService.startOAuth(scope);
  };

  const triggerRender = (): void => {
    gitHubReposAdapter.triggerRender();
  };

  return (
    <GenericRightPaneComponent {...genericPaneProps}>
      <div className="panelFormWrapper">
        <div className="panelMainContent">
          <GitHubReposComponent {...gitHubReposProps} />
        </div>
      </div>
    </GenericRightPaneComponent>
  );
};
