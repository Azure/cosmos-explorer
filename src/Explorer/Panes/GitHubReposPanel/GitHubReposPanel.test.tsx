import { shallow } from "enzyme";
import React from "react";
import { GitHubClient } from "../../../GitHub/GitHubClient";
import { JunoClient } from "../../../Juno/JunoClient";
import Explorer from "../../Explorer";
import { GitHubReposPanel } from "./GitHubReposPanel";
const props = {
  explorer: new Explorer(),
  closePanel: (): void => undefined,
  gitHubClientProp: new GitHubClient((): void => undefined),
  junoClientProp: new JunoClient(),
  openNotificationConsole: (): void => undefined,
};
describe("GitHub Repos Panel", () => {
  it("should render Default properly", () => {
    const wrapper = shallow(<GitHubReposPanel {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
