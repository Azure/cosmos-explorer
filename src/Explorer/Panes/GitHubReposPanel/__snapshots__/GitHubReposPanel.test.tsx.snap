// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub Repos Panel should render Default properly 1`] = `
<form
  className="panelFormWrapper"
>
  <div
    className="panelMainContent"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <GitHubReposComponent
      addRepoProps={
        Object {
          "container": Explorer {
            "_isInitializingNotebooks": false,
            "_resetNotebookWorkspace": [Function],
            "canSaveQueries": [Function],
            "collapsedResourceTreeWidth": 36,
            "commandBarComponentAdapter": CommandBarComponentAdapter {
              "container": [Circular],
              "isNotebookTabActive": [Function],
              "parameters": [Function],
              "tabsButtons": Array [],
            },
            "databases": [Function],
            "isAccountReady": [Function],
            "isFixedCollectionWithSharedThroughputSupported": [Function],
            "isHostedDataExplorerEnabled": [Function],
            "isNotebookEnabled": [Function],
            "isNotebooksEnabledForAccount": [Function],
            "isResourceTokenCollectionNodeSelected": [Function],
            "isSchemaEnabled": [Function],
            "isServerlessEnabled": [Function],
            "isShellEnabled": [Function],
            "isSynapseLinkUpdating": [Function],
            "isTabsContentExpanded": [Function],
            "memoryUsageInfo": [Function],
            "notebookBasePath": [Function],
            "notebookServerInfo": [Function],
            "onRefreshDatabasesKeyPress": [Function],
            "onRefreshResourcesClick": [Function],
            "provideFeedbackEmail": [Function],
            "queriesClient": QueriesClient {
              "container": [Circular],
            },
            "refreshNotebookList": [Function],
            "resourceTokenCollection": [Function],
            "resourceTokenCollectionId": [Function],
            "resourceTokenDatabaseId": [Function],
            "resourceTokenPartitionKey": [Function],
            "resourceTree": ResourceTreeAdapter {
              "container": [Circular],
              "copyNotebook": [Function],
              "databaseCollectionIdMap": Map {},
              "koSubsCollectionIdMap": Map {},
              "koSubsDatabaseIdMap": Map {},
              "parameters": [Function],
            },
            "resourceTreeForResourceToken": ResourceTreeAdapterForResourceToken {
              "container": [Circular],
              "parameters": [Function],
            },
            "selectedDatabaseId": [Function],
            "selectedNode": [Function],
            "setInProgressConsoleDataIdToBeDeleted": undefined,
            "setNotificationConsoleData": undefined,
            "sparkClusterConnectionInfo": [Function],
            "splitter": Splitter {
              "bounds": Object {
                "max": 400,
                "min": 240,
              },
              "direction": "vertical",
              "isCollapsed": [Function],
              "leftSideId": "resourcetree",
              "onResizeStart": [Function],
              "onResizeStop": [Function],
              "splitterId": "h_splitter1",
            },
            "tabsManager": TabsManager {
              "activeTab": [Function],
              "openedTabs": [Function],
            },
          },
          "getRepo": [Function],
          "pinRepo": [Function],
        }
      }
      authorizeAccessProps={
        Object {
          "authorizeAccess": [Function],
          "scope": "public_repo",
        }
      }
      onCancelClick={[Function]}
      onOkClick={[Function]}
      reposListProps={
        Object {
          "branchesProps": Object {},
          "pinRepo": [Function],
          "pinnedReposProps": Object {
            "repos": Array [],
          },
          "unpinRepo": [Function],
          "unpinnedReposProps": Object {
            "hasMore": true,
            "isLoading": true,
            "loadMore": [Function],
            "repos": Array [],
          },
        }
      }
      resetConnection={[Function]}
      showAuthorizeAccess={true}
    />
  </div>
</form>
`;
